### PR TITLE
fix: change Tag close icon from span to button for keyboard accessibility (#57522)

### LINKED ISSUE #57522: Tag: close icon cannot be machine read and or used by keyboard
### Reproduction link

[Edit on StackBlitz](https://stackblitz.com/edit/react-2ntxvwnz?file=demo.tsx)

### Steps to reproduce

1. Render <Tag closable>Example</Tag>
2. Press Tab to try to focus the close icon
3. Inspect the close icon element in DevTools

### Expected behavior

I expect to be able to quickly navigate a webpage using the keyboard.

- Close icon is focusable via Tab
- Close icon can be actioned via Enter or Space
- Close icon has an accessible name (e.g. aria-label) for screen readers
- Visible focus indicator appears on the close icon when focused

### Current behavior

The close icon renders as:

```html
<span class="ant-tag-close-icon">
  <span class="anticon anticon-close">...</span>
</span>
```

This <span> is not in the tab order and has no semantic role or label. Keyboard users cannot reach or activate it.

| Environment | Info |
| --- | --- |
| antd | 6.3.5 |
| React | 18.x |
| System | All |
| Browser | All |

---
### Context

This is a WCAG 2.1 Level A failure — specifically [SC 2.1.1 Keyboard](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html) and [SC 4.1.2 Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-
role-value.html). Any interactive element must be operable via keyboard and expose its name and role to assistive technology.

### PR DESCRIPTION
**Pull request by tyreer**
### 🤔 This is a ...

- [x] 🐞 Bug fix
- [x] ⌨️ Accessibility improvement

### 🔗 Related Issues

close #57522

### 💡 Background and Solution

When Tag has `closable` set, the close icon is rendered as a `<span>` with an `onClick` handler. This is not keyboard-focusable, has no implicit ARIA role, and cannot be activated via Enter or Space. This is a WCAG 2.1 Level A failure (SC 2.1.1 Keyboard, SC 4.1.2 Name, Role, Value).

**Root cause:** Tag's `closeIconRender` used `replaceElement()`, which short-circuits when the icon is a valid React element (like `<CloseOutlined />`). It clones the icon's props instead of using the wrapper, so the close icon was always rendered as a `<span>` regardless.

**Solution:**
- Changed the close icon wrapper from `<span>` to `<button type="button">` in Tag's `closeIconRender`
- Added guard to avoid nested `<button>` when a user passes a custom `<button>` as `closeIcon`
- Added button reset styles and `:focus-visible` outline on `.ant-tag-close-icon`
- Close button is properly `disabled` when Tag is disabled
- Icon inside close button marked `aria-hidden` to prevent redundant screen reader announcements
- Kept the shared `useClosable` hook fallback as `<span>` to avoid nested buttons in Notification/Modal/Drawer/Tour (those components have their own button wrappers)

No API changes. The `aria-label` is already provided via the existing locale system.

d0c1d5c250192801a544ec1a905482d41dc7d50d1ee763c1886c9c0e00595742

### 📝 Change Log

| Language | Changelog |
| --- | --- |
| 🇺🇸 English | ⌨️ Fix Tag close icon not being keyboard-accessible by changing wrapper from `span` to `button`. |
| 🇨🇳 Chinese | ⌨️ 修复 Tag 关闭图标无法通过键盘访问的问题，将包裹元素从 `span` 改为 `button`。 |

### CODE REVIEW

File: components/tag/index.tsx [hunk 1]
```diff
@@ -180,18 +180,26 @@ const InternalTag = React.forwardRef<HTMLSpanElement | HTMLAnchorElement, TagPro
...
+          });
+        }
+        return (
+          <button
+            type="button"
+            className={`${prefixCls}-close-icon`}
+            onClick={handleCloseClick}
+            tabIndex={0}
+          >
             {iconNode}
```
[the remark this answers is not in the document]
Comment (tyreer) [reply components/tag/index.tsx]: Good catch — addressed. The icon inside the close button now gets aria-hidden="true" so screen readers only announce the button's aria-label from the locale system.

File: components/tag/index.tsx [hunk 2]
```diff
@@ -180,18 +180,26 @@ const InternalTag = React.forwardRef<HTMLSpanElement | HTMLAnchorElement, TagPro
...
+            },
+            className: clsx(iconNode.props?.className, `${prefixCls}-close-icon`),
+          });
+        }
+        return (
+          <button
+            type="button"
+            className={`${prefixCls}-close-icon`}
+            onClick={handleCloseClick}
+            tabIndex={0}
```
[the remark this answers is not in the document]
Comment (tyreer) [reply components/tag/index.tsx]: Good catch — addressed. Reverted the useClosable fallback wrapper back to <span>. Tag handles its own <button> wrapping via closeIconRender, so the shared hook doesn't need to change. Confirmed Drawer/
Notification snapshot tests still pass.

File: components/_util/hooks/useClosable.tsx [hunk 3]
```diff
@@ -117,9 +117,9 @@ const computeCloseIcon = (
         ...ariaOrDataProps,
       })
     ) : (
-      <span aria-label={closeLabel} {...ariaOrDataProps}>
+      <button type="button" aria-label={closeLabel} {...ariaOrDataProps}>
         {finalCloseIcon}
-      </span>
+      </button>
```
[the remark this answers is not in the document]
Comment (tyreer) [reply components/_util/hooks/useClosable.tsx]: Valid concern. However, no existing demo or test combines href with closable on the same Tag, and this edge case wasn't accessible before this PR either. I've noted it as a known limitation to address in a follow-up if needed.

File: components/tag/index.tsx [hunk 4]
```diff
@@ -180,18 +180,28 @@ const InternalTag = React.forwardRef<HTMLSpanElement | HTMLAnchorElement, TagPro
...
+          ? React.cloneElement<any>(iconNode, { 'aria-hidden': true })
+          : iconNode;
+        return (
+          <button
+            type="button"
+            className={`${prefixCls}-close-icon`}
+            onClick={handleCloseClick}
+          >
+            {decorativeIcon}
+          </button>
```
[the remark this answers is not in the document]
Comment (tyreer) [reply components/tag/index.tsx:206]: [358020c](https://github.com/ant-design/ant-design/pull/57525/commits/358020c93761e24b51ef63a14a60be464789ace3)

File: components/tag/index.tsx [hunk 5]
```diff
@@ -180,18 +180,28 @@ const InternalTag = React.forwardRef<HTMLSpanElement | HTMLAnchorElement, TagPro
...
         );
```
[the remark this answers is not in the document]
Comment (tyreer) [reply components/tag/index.tsx:207]: 9680bf2

File: components/tag/index.tsx [hunk 6]
```diff
@@ -180,18 +180,31 @@ const InternalTag = React.forwardRef<HTMLSpanElement | HTMLAnchorElement, TagPro
     const [, mergedCloseIcon] = useClosable(pickClosable(props), pickClosable(tagContext), {
       closable: false,
       closeIconRender: (iconNode: React.ReactNode) => {
-        const replacement = (
-          <span className={`${prefixCls}-close-icon`} onClick={handleCloseClick}>
-            {iconNode}
-          </span>
+        // If the custom closeIcon is already an interactive element, merge props instead of wrapping
+        if (React.isValidElement<any>(iconNode) && iconNode.type === 'button') {
+          return React.cloneElement<any>(iconNode, {
```
[the remark this answers is not in the document]
Comment (tyreer) [reply components/tag/index.tsx:185]: The check intentionally only handles native <button> elements — detecting component-based buttons via displayName is fragile and no existing demo or test passes a component-based closeIcon. The closeIcon API is designed for icons (e.g. <CloseOutlined />), not full Button components. If a real user reports this, we can address it in a follow-up.

**Comment by tyreer** [comment]
Here's how I'm making sense of the automated input.

All CI checks pass (lint, unit tests across React 18/19, build, node tests, coverage).

The checks summary shows some entries marked "Failing" or "Cancelled," but these are stale results from earlier force-pushes during iteration. The most recent run for each check shows ✅. The "Visual Regression Diff Wait Approve" gate is a manual approval step awaiting maintainer review, not a test failure.

Visual regression report (25 diffs):

- **Layout diffs (15):** False positives — the layout demos use new Date().getFullYear() in their footer copyright, which produces a different year than the base snapshots' mocked date. Unrelated to this PR.
- **Tag diffs:** Expected — the <span> → <button> change for the close icon intentionally alters the rendered DOM, so pixel diffs are correct and reflect the purpose of this PR.
- **Splitter diff:** Unrelated to this PR's changes.
- **Bundle size report:** The -64% figure is misleading — it reflects a minified vs unminified artifact naming change on master, not an actual size regression. The actual footprint of this change is ~6.25kB in
components/tag/style/index.ts.

This PR is ready for maintainer review as far as I can tell.

**Comment by afc163** [comment]
See https://github.com/ant-design/ant-design/actions/runs/23855245107/job/69603480196?pr=57525

**Comment by tyreer** [comment]
@afc163 Reverted the unrelated drawer snapshot in 9a07ea9 — that was a timezone-sensitive ant-picker-cell-today date drift from the merge, not related to the Tag changes. The pull_request CI workflows appear to
need re-approval after the force-push. Could you approve the pending runs?

**Comment by tyreer** [comment]
@afc163 Summary of what I'm seeing in the PR:

- All CI checks pass on the latest commit (ebebd15)
- The visual regression diffs (10 changed, 3 removed) are either expected Tag diffs from the `span` → `button` change, or unrelated noise (cascader-ellipsis-debug removals, alert-error-boundary)
- The snapshot test failure you flagged has been resolved (drawer snapshot reverted)
- Bot comments are all duplicates of the same two items: `href` + `closable` nesting (pre-existing, out of scope for this PR) and `any` type usage (style nitpick)

Is there anything else needed from our side, or is this ready for your review?

File: components/tag/style/index.ts [hunk 7]
```diff
@@ -77,9 +77,25 @@ const genBaseStyle: GenerateStyle<TagToken, CSSInterpolation> = (token) => {
...
+        // Reset button styles
+        display: 'inline-flex',
+        alignItems: 'center',
+        justifyContent: 'center',
+        padding: 0,
+        border: 'none',
+        background: 'transparent',
+        lineHeight: 0,
+        verticalAlign: '-0.125em',
+
```
Comment (KittyGiraudel) [inline components/tag/style/index.ts:89]: Maybe we want to do `font: inherit` as well?

File: components/tag/style/index.ts [hunk 8]
```diff
@@ -77,9 +77,25 @@ const genBaseStyle: GenerateStyle<TagToken, CSSInterpolation> = (token) => {
...
         transition: `all ${token.motionDurationMid}`,
 
+        // Reset button styles
+        display: 'inline-flex',
+        alignItems: 'center',
+        justifyContent: 'center',
+        padding: 0,
+        border: 'none',
+        background: 'transparent',
+        lineHeight: 0,
```
Comment (KittyGiraudel) [inline components/tag/style/index.ts:87]: I think `lineHeight: 1` would make more sense and would achieve the same thing.

File: components/tag/style/index.ts [hunk 9]
```diff
@@ -77,9 +77,25 @@ const genBaseStyle: GenerateStyle<TagToken, CSSInterpolation> = (token) => {
...
+        verticalAlign: '-0.125em',
+
         '&:hover': {
           color: token.colorTextHeading,
         },
+
+        '&:focus-visible': {
+          outline: `${unit(token.lineWidthFocus)} solid ${token.colorPrimaryBorder}`,
+          outlineOffset: 1,
```
Comment (KittyGiraudel) [inline components/tag/style/index.ts:96]: Can we use the `genFocusStyle` utility instead?

File: components/tag/style/index.ts [hunk 10]
```diff
@@ -77,9 +77,25 @@ const genBaseStyle: GenerateStyle<TagToken, CSSInterpolation> = (token) => {
...
+          borderRadius: token.borderRadiusXS,
```
Comment (KittyGiraudel) [inline components/tag/style/index.ts:97]: This can be defined on the normal state of the button, it doesn’t have to be defined when focused. :)

File: components/tag/index.tsx [hunk 11]
```diff
@@ -180,18 +180,31 @@ const InternalTag = React.forwardRef<HTMLSpanElement | HTMLAnchorElement, TagPro
...
+        }
+        const decorativeIcon = React.isValidElement<any>(iconNode)
+          ? React.cloneElement<any>(iconNode, { 'aria-hidden': true })
+          : iconNode;
+        return (
+          <button
+            type="button"
+            className={`${prefixCls}-close-icon`}
+            onClick={handleCloseClick}
+            disabled={mergedDisabled}
```
[the remark this answers is not in the document]
Comment (KittyGiraudel) [reply components/tag/index.tsx:203]: If we get a `href` prop, we shouldn’t render a `<button>` element inside it because it will produce invalid HTML, good catch.. In that case, going the `span` + `role="button"` + focus/keyboard handler makes sense. 

But maybe to get this PR out of the door and greatly improve accessibility for most cases, we could implement the changes *only* when we don’t get a `href` prop?

**Comment by afc163** [comment]
图片

**Comment by tyreer** [comment]
Looks like https://github.com/ant-design/ant-design/pull/58067 has addressed this, making this PR redundant.

Thanks for the back and forth and review. Glad to see the accessibility work done!
